pkg/controller/acmechallenges/sync.go

```diff
@@ -131,6 +133,26 @@ func (c *Controller) Sync(ctx context.Context, ch *cmapi.Challenge) (err error)
 		return nil
 	}
 
+	// check for CAA records.
+	// CAA records are static, so we don't have to present anything
+	// before we check for them.
+
+	// Find out which identity the ACME server says it will use.
+	dir, err := cl.Discover(ctx)
+	if err != nil {
+		return err
+	}
+	// TODO(dmo): figure out if missing CAA identity in directory
+	// means no CAA check is performed by ACME server or if any valid
+	// CAA would stop issuance (strongly suspect the former)
+	if len(dir.CAA) != 0 {
+		err := dnsutil.ValidateCAA(ch.Spec.DNSName, dir.CAA, ch.Spec.Wildcard)
+		if err != nil {
+			ch.Status.Reason = fmt.Sprintf("CAA self-check failed: %s", err)
+			return err
```

Comment:
Returning an error here will trigger the controllers back-off for failing syncs (exponential with a base of 10s and a cap of 30m).

Do we want this behaviour? The self check usually waits a fixed 10s before attempts right now for regular DNS/HTTP self check failures: https://github.com/jetstack/cert-manager/blob/ec1305cfd638f0972be19f7a624358ec7e3846ce/pkg/controller/acmechallenges/sync.go#L183

Reply:
I think we do. While usual self-checks can fix themselves (propagation delays), CAA failures tend to be more permanent (wrong issuer listed)